test/functional/p2p_eviction.py

```diff
@@ -0,0 +1,136 @@
+        # (32 - 10 outbound - 1 feeler). 20 inbound peers are protected from eviction:
+        # 4 by netgroup, 4 that sent us blocks, 4 that sent us transactions and 8 via lowest ping time
+        self.extra_args = [['-maxconnections=32']]
+
+    def run_test(self):
+        self.protected_peers = set() # peers that we expect to be protected from eviction
+        current_peer = -1
+        node = self.nodes[0]
+        node.generatetoaddress(101, node.get_deterministic_priv_key().address)
+
+        self.log.info("Create 4 peers and protect them from eviction by sending us a block")
+        for _ in range(4):
+            block_peer = node.add_p2p_connection(SlowP2PDataStore())
+            current_peer += 1
+            block_peer.sync_with_ping(timeout=5)
+            best_block = node.getbestblockhash()
+            tip = int(best_block, 16)
+            best_block_time = node.getblock(best_block)['time']
+            block = create_block(tip, create_coinbase(node.getblockcount() + 1), best_block_time + 1)
+            block.solve()
+            block_peer.send_blocks_and_test([block], node, success=True)
+            self.protected_peers.add(current_peer)
+
+        self.log.info("Create 5 slow-pinging peers, making them eviction candidates")
+        for _ in range(5):
+            node.add_p2p_connection(SlowP2PInterface())
+            current_peer += 1
+
+        self.log.info("Create 4 peers and protect them from eviction by sending us a tx")
+        for i in range(4):
+            txpeer = node.add_p2p_connection(SlowP2PInterface())
+            current_peer += 1
+            txpeer.sync_with_ping(timeout=5)
+
+            prevtx = node.getblock(node.getblockhash(i+1), 2)['tx'][0]
+            rawtx = node.createrawtransaction(
+                inputs=[{'txid': prevtx['txid'], 'vout': 0}],
+                outputs=[{node.get_deterministic_priv_key().address: 50 - 0.00125}],
+            )
+            sigtx = node.signrawtransactionwithkey(
+                hexstring=rawtx,
+                privkeys=[node.get_deterministic_priv_key().key],
+                prevtxs=[{
+                    'txid': prevtx['txid'],
+                    'vout': 0,
+                    'scriptPubKey': prevtx['vout'][0]['scriptPubKey']['hex'],
+                }],
+            )['hex']
+            txpeer.send_message(msg_tx(FromHex(CTransaction(), sigtx)))
+            self.protected_peers.add(current_peer)
+
+        self.log.info("Create 8 peers and protect them from eviction by having faster pings")
+        for _ in range(8):
+            fastpeer = node.add_p2p_connection(P2PInterface())
+            current_peer += 1
+            wait_until(lambda: "ping" in fastpeer.last_message, timeout=10)
+
+        # Make sure by asking the node what the actual min pings are
+        peerinfo = node.getpeerinfo()
+        pings = [[0 for i in range(2)] for j in range(len(node.p2ps))]
```

Comment:
Using a list of lists like this is a slightly confusing structure. Consider using a dictionary, and then sorting the `dict.items()` items by value.

Reply:
Changed it to a dictionary.